### PR TITLE
Add fuzz testing for UTF8 LIKE pruning

### PR DESCRIPTION
**Pull request by alamb**
Draft as it builds on https://github.com/apache/datafusion/pull/12978

## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/507

## Rationale for this change

While working on https://github.com/apache/datafusion/pull/12978 with @adriangb and @findepi  I am having nightmares of subtle bugs introduced with truncated statistics

## What changes are included in this PR?

Fuzz tests for pruning with truncated statistics / prefix values

## Are these changes tested?
It is only tests
```shell
cargo test --test fuzz -- pruning
```

## Are there any user-facing changes?

No, tests only

### CODE REVIEW

File: datafusion/core/tests/fuzz_cases/pruning.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+use rand::{thread_rng, Rng};
+use std::collections::HashSet;
+use std::sync::Arc;
+
+/// Tests for `LIKE` with truncated statistics to validate incrementing logic
+///
+/// Create several 2 row batches and ensure that `LIKE` with the min and max value
+/// are correctly pruned even when the "statistics" are trunated.
+#[test]
+fn test_prune_like_truncated_statistics() {
```
Comment (adriangb) [inline datafusion/core/tests/fuzz_cases/pruning.rs:34]: I think it's also worth having tests for `=` and maybe other operators, it was not immediately obvious to me that there wasn't a bug with those as well.

File: datafusion/core/tests/fuzz_cases/pruning.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    // Make 2 row random UTF-8 strings
+    let mut rng = thread_rng();
+    let statistics = TestPruningStatistics::new(&mut rng, 100);
```
Comment (adriangb) [inline datafusion/core/tests/fuzz_cases/pruning.rs:37]: I imagine a lot of the bugs are going to be around edge cases: empty strings, non-ascii characters, etc. Is there any way we could inject those into the randomness? Maybe what we need here more than random fuzzing is a matrix style test:
- Generate N full length values, including some random ones?
- Arrange them into row groups in multiple orders, of multiple sizes
- Truncate the stats to lengths between 1 and large

And make sure the results with and without pruning match?

Comment (alamb) [reply datafusion/core/tests/fuzz_cases/pruning.rs:37]: THis is a good idea
